### PR TITLE
avoid record streams and make color printing a config

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #195

1. Avoid `recordStream ` to partly resolve issue #191 (the other part needs to be fixed in pytorch).
2. Color printing on aws/mast doesn't render well, thus making it into a config which defaults to "disabled" in `config_manager.py`. Enabling it in `debug_model.py` by default.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -178,7 +174,7 @@ def main(job_config: JobConfig):
     num_flop_per_token = get_num_flop_per_token(
         model_param_count, model_config, job_config.training.seq_len
     )
-    if _is_local_logging:
+    if job_config.misc.enable_color_printing:
```
Comment (gnadathur) [inline train.py]: Should we just teach the `Color` class to nullify color if this flag is set. That way we can still have the print but in normal mode. From this change it looks like we are loosing the log message if color printing is not enabled ? Maybe we can do this in a subsequent change.
Wondering if we can something like this to the `Color` class:

```
@classmethod
def enabled(cls, is_enabled: bool):
     if is_enabled is False:
           cls.black = ..... = ""   #set every color prefix to empty string
```
Then in setup.py:

`
Color.enabled(job_config.misc.enable_color_printing)
`

This way we can retain the log message but just nullify color prefix. We also dont need the `if` check. Would this work ?

File: train_configs/debug_model.toml [hunk 2]
```diff
@@ -46,3 +46,6 @@ dataset = "alpaca"   # supported datasets = alpaca (52K), minipile (1M), c4 (177
 [activation_checkpoint]
 mode = 'selective'  # ['none', 'full', 'selective']
 selective_ac_option = '2'  # 'int' = ac every positive int layer or 'op', ac based on ops policy
+
+[misc]
+enable_color_printing = true
```
Comment (gnadathur) [inline train_configs/debug_model.toml]: Should we control this flag automatically based on scheduler ? So for MAST and SLURM setup does the right thing.

File: torchtrain/utils.py [hunk 3]
```diff
@@ -87,6 +87,10 @@ def init_distributed(job_config):
         "nccl", timeout=timedelta(seconds=job_config.comm.init_timeout_seconds)
     )
 
+    # to mitigate the memory issue that collective tensors
+    # in DTensor holding memory longer than they should
```
Comment (wanchaol) [inline torchtrain/utils.py]: nit: I think this is more related to: collectives that uses `async_op=True`, not specifically to DTensor

File: torchtrain/config_manager.py [hunk 4]
```diff
@@ -278,6 +280,14 @@ def __init__(self):
             help="Flight recorder ring buffer size, >0 means recording by default, 0 means disabled",
         )
 
+        # misc settings
+        self.parser.add_argument(
+            "--misc.enable_color_printing",
```
Comment (wanchaol) [inline torchtrain/config_manager.py]: should this be part of `metrics` option, instead of adding a new `misc` section?

File: train.py (hunk 1 above)
Comment (tianyu-l) [reply train.py]: added another dataclass NoColor, to make code logically simpler

File: train_configs/debug_model.toml (hunk 2 above)
Comment (tianyu-l) [reply train_configs/debug_model.toml]: agree, but not sure what's the most elegant way of detecting this...

File: torchtrain/utils.py [hunk 5]
```diff
@@ -87,6 +87,10 @@ def init_distributed(job_config):
         "nccl", timeout=timedelta(seconds=job_config.comm.init_timeout_seconds)
     )
 
+    # to mitigate the memory issue that collectives using
+    # async_op=True holds memory longer than they should
```
Comment (awgu) [inline torchtrain/utils.py]: maybe mention TP as an example?
```suggestion
    # to mitigate the memory issue that collectives using
    # async_op=True hold memory longer than they should
    # such as those in tensor parallelism
```
